crates/core/transaction/src/plan/build.rs

```diff
@@ -188,273 +153,56 @@ impl TransactionPlan {
+    /// Returns: `Transaction`
+    pub fn authorize_with_auth<R: CryptoRng + RngCore + Debug>(
+        &self,
+        rng: &mut R,
+        auth_data: &AuthorizationData,
+        mut transaction: Transaction,
+    ) -> Result<Transaction> {
+        // Do some basic input sanity-checking.
+        let spend_count = transaction.spends().count();
+        if auth_data.spend_auths.len() != spend_count {
+            anyhow::bail!(
+                "expected {} spend auths but got {}",
+                spend_count,
+                auth_data.spend_auths.len()
+            );
         }
 
-        // We build the actions sorted by type, with all spends first, then all
-        // outputs, etc.  This order has to align with the ordering in
-        // TransactionPlan::effect_hash, which computes the auth hash of the
-        // transaction we'll build here without actually building it.
-
-        // Start building the transaction's spends.
-        let mut in_progress_spend_actions = Vec::new();
-        for spend_plan in self.spend_plans().cloned() {
-            let note_commitment = spend_plan.note.commit();
-            let auth_path = witness_data
-                .state_commitment_proofs
-                .get(&note_commitment)
-                .context(format!("could not get proof for {note_commitment:?}"))?
-                .clone();
+        // Derive the blinding factors from `TransactionPlan``.
+        let mut synthetic_blinding_factor = Fr::zero();
 
+        for spend_plan in self.spend_plans() {
             synthetic_blinding_factor += spend_plan.value_blinding;
-            let fvk_ = fvk.clone();
-            in_progress_spend_actions.push(tokio::spawn(async move {
-                //Add dummy auth sig for UnauthTransaction
-                let auth_sig = [0; 64].into();
-                spend_plan.spend(&fvk_, auth_sig, auth_path, witness_data.anchor)
-            }));
         }
-
-        // Start building the transaction's outputs.
-        let mut in_progress_output_actions = Vec::new();
-        let dummy_payload_key: PayloadKey = [0u8; 32].into();
-        // If the memo_key is None, then there is no memo, and we populate the memo key
-        // field with a dummy key.
-        for output_plan in self.output_plans().cloned() {
-            // Outputs subtract from the transaction's value balance.
+        for output_plan in self.output_plans() {
             synthetic_blinding_factor += output_plan.value_blinding;
-            let ovk = fvk.outgoing().clone();
-            let memo_key = memo_key.as_ref().unwrap_or(&dummy_payload_key).clone();
-            in_progress_output_actions.push(tokio::spawn(async move {
-                output_plan.output(&ovk, &memo_key)
-            }));
         }
-
-        // Start building the transaction's swaps.
-        let mut in_progress_swap_actions = Vec::new();
-        for swap_plan in self.swap_plans().cloned() {
+        for swap_plan in self.swap_plans() {
```

Comment:
We have a lot of loops over each type of action here. It would be easy to miss some. What if instead we added a method `ActionPlan::value_blinding()` whose body was a `match` over the possible kinds of `ActionPlan`, returning `Fr::zero()` for action plans that don't have a value blinding (because their effect on the value balance is public).

Then this could just be
```
for action_plan in &self.action_plans {
    synthetic_blinding_factor += action_plan.value_blinding();
}
```

Reply:
That's a great idea, I switched to this pattern!